### ISSUE TITLE
vDOMSON a different way to define vDOM

### ISSUE THREAD

**Issue by aronallen** [snabbdom/snabbdom#186]
Hi,

We are currently experimenting with using WebWorkers to create a vDOM tree, and we thought it could be relevant to rethink how we create vDOM elements.

Basically what we have discovered is that we could represent any vDOM tree with a subset of JSON. We could define this subset as vDOMSON, besides providing a simple way to Stringify a vDOM structure, it could also serve as a base for how all vDOMs are represented, so the vDOM declarations are independant of the vDOM library.

Right now we write vDOM like this, and it is fairly consistent across vDOM libraries.

``` javascript
h(
  'div', 
  {
    attributes : {class : 'hello'}
  }, 
  [
    h('span', {}, 'hello'), 
   '!'
  ]
)
```

What we suggest is to use vDOMSON and pipe that through the vDOM lib.

``` javascript
snabbdomify([
  'div', 
  {
    attributes : {class : 'hello'}
  }, 
  [
    ['span', {}, 'hello'], 
   '!'
  ]
])
```

I believe having this approach could increase decoupling so vDOM libs are easily exchangeable.
Taking this further we could pass vDOMSON directly into the patch function.

**Comment by developit** [comment]
I tried this in an experiment recently and ran into an issue: it's impossible to determine if entries within the `children` array are VDOM elements or nested children:

```js
[
  'div',
  { foo: 'bar' },
  [
    ['span', {}, 'hello'],   // is this a span, or some text nodes?
    'div'
  ]
]
```

**Comment by mightyiam** [comment]
@developit "some text nodes" doesn't make sense to me. An array inside the children array is an element. Text nodes must be just text. Am I missing something?

**Comment by developit** [comment]
Arrays of children can be infinitely nested, at least that is the case with VDOM generated from JSX. Is that not true for snabbdom?

Here's some JSX that produces nested children:

```js
<div>a { [1,2,3].map( n => <span>{n}</span> ) }</div>
// produces:
h('div', null, 'a', [h('span', null, 1), h('span', null, 2), h('span', null, 3)]);
```

**Comment by mightyiam** [comment]
I love the idea of using primitives to define trees. I'm thinking that this is what we're already doing behind the scenes. I mean, `h` returns a tree of primitives. Why even use `h`? Yet, `h` does [do some more things](https://github.com/snabbdom/snabbdom/blob/master/src/h.ts#L19-L41) that  I've not gotten into yet. So I don't know. I bet @paldepind and @staltz know well.

**Comment by mightyiam** [comment]
```js
[
  'div',
  { foo: 'bar' },
  [ // this is the children array of the root element
    ['span', {}, 'hello'],   // this can't be text nodes. An array here can only be an element.
    ['span', null, 'hello'],   // this, as well. Can't be text nodes. Because it is an array.
    'div' // this seems like a text node to me.
  ]
]
```

This is how I see it.

Text nodes are only strings directly inside the children array.

**Comment by developit** [comment]
This means you cannot use this format with JSX and must apply a flattening step when converting.

Why not use an object with properties to represent each VNode?

```js
{
  nodeName: 'div',
  attributes: { foo: 'bar' },
  children: [
    { nodeName: 'span', children: ['hello'] },
    { nodeName: 'div' }
  ]
}
```

To me this seems like the most direct translation, given that most frameworks can actually use these POJO's unmodified ([random example](https://jsfiddle.net/pcfq3pq8/)).

**Comment by aronallen** [comment]
We are already using the array format in a project. It maps pretty well to snabbdoms api.
Of course it we want to standardize this across vDOM libraries, we have to consider that each lib might have individual needs. I prefer the array format over the object format, simply because it is less verbose.

I wrote a small funciton i embeded below, to show how one might transpose vDOMSON to snabbdom.

```typescript
export const vDOMSONtoSnabbdom = (node: vDOMSON) => {
    const [tagname, options, children] = node;
    if (Array.isArray(children)) {
        return h(tagname, options || {}, children.map(child => typeof child === 'string' ? child : vDOMSONtoSnabbdom(child)))
    } else {
        return h(tagname, options || {}, children);
    }
}
```

@developit your jsx example would be written like this in vDOMSON

```javascript
['div', null, [
  'a',
  ['span', null, '1'], 
  ['span', null, '2'],
  ['span', null, '3']
]];
```

**Comment by forceuser** [comment]
My proposition:
```json
["div", null, [
    ["span", {"key": 1}, "span node"],
    [null, {"key": 2}, "text node with key"],
    ["!", {"key": 3}, "comment node with key"]
]]
```

**Comment by mightyiam** [comment]
Why does the `h` function exist in the first place? Doesn't it do something useful other than blurting out almost the same input that it is given?

**Comment by developit** [comment]
Generally `h()` does [normalization](https://github.com/developit/preact/blob/master/src/h.js#L17-L50). It's also a way to hook VNode creation. In some cases, VNode objects contain more data than described here, like React's `$$typeof` hint or Preact's VNode constructor.

**Comment by mightyiam** [comment]
Can this normalization and hook stuff happen at a different layer? I mean, can we get rid of `h`?

**Comment by developit** [comment]
@mightyiam No, that would mean the VDOM renderer would have to deal with a whole bunch more cases internally.  I think that's actually part of the reason this format has become so popular: you define the arguments to a library-provided function that can then do *anything* with them.  If JSX/hyperscript/compilers/whatever instead output some very specific format like the ones described in this thread (strict array syntax or named properties, either are very specific), we'd lose a lot of that versatility.

As an example, you can directly render JSX/hyperscript to HTML with no VDOM intermediary.  You might find the source of [vhtml](https://npm.im/vhtml) an interesting study in alternative output targets for this tiny, versitile convention.  You could have `h()` assemble a DOM, XML strings, pre-serialized JSON, instances of specific types, etc.  It's open-ended, which makes it a good standard.

TL;DR: Raw JSON can only ever be the specific raw JSON you define. `h(args)` can output anything.

**Comment by mightyiam** [comment]
@developit also, `h()` allows some features. For example, the first argument is not only a tag. It can be `div.foo#bar`. But I've never used that. I feel that `h()` is not going anywhere...

I do not see `h()` as being a standard, because there are so many implementations with different syntaxes, right?

In any case, the conversation here is about standardization of the JSON *output* of `h()`.

In that regard, I have something to add to the conversation and that is #207.

**Comment by forceuser** [comment]
If not think of VDOMSON only in context of snabbdom, i am also would like to add format for fragments

```js
["div", null, [
    ["h3", {key: "header"}, "text"],
    [null, {key: "fragment1"}, [ // fragment
        [null, null, "text node"],
        ["!", {key: "comment 1"}, "comment"],
        ["span", null, "span content"]
    ],
    ["div", null, null] // empty div
]]
```

And i think VDOMSON should be strict: every node should be described by array and all 3 parameters should be specified, even if it is null

**Comment by mightyiam** [comment]
@forceuser does this have to do with `DocumentFragment` or what kind of a fragment is this? I'm unfamiliar.

**Comment by hville** [comment]
Have you looked at the old [jsonml](http://www.jsonml.org) format? It is very similar.

**Comment by aronallen** [comment]
I found a way to use vDOMSON with JSX, it is quite simple actually.

```typescript
export const JSX = {
  createElement (tagName: string, options = {}, ...children): vDOMSON {
    return [tagName, options, children];
  },
  __spread: (a) => Object.assign({}, ...a)
}
```

**Comment by localvoid** [comment]
@aronallen Here are some problems with your vDOMSON:

1. Inefficient data structure
2. Doesn't describe normalization process, many libraries are handling positional and nested indexing differently and there are no ideal solutions, all solutions have pros and cons
3. Doesn't have xss protection ( http://danlec.com/blog/xss-via-a-spoofed-react-element )

And btw, try this with your jsx implementation `<div>span{null}123</div>` ;)

**Comment by aronallen** [comment]
@localvoid thanks for the feedback.
1. How is the datastructure ineffecient?
2. Can you eloborate this, I don't quite follow
3. I think I understand the attack, but how could it work here? There is no such thing as dangerouslySetInnerHTML in snabbdom.

```<div>span{null}123</div>``` throws ```Cannot read property 'length' of null``` in the WebWorker thread

**Comment by aronallen** [comment]
@localvoid I actually had a more recent JSX parser
```typescript
export const JSX = {
    createElement(tagName: string, options, ...children): vDOMSON {
        const result = (children || []).reduce((acc, n) => {
            // skip empty arrays and strings
            if (n === null || n === undefined) {
                return acc;
            } else if (n.length === 0) {
                return acc;
            } else if (n && n[0] && n[0].map) {
                return acc.concat(n);
            } else {
                return [...acc, n];
            }
        }, []);
        return [tagName, options, result];
    },
    __spread: (a) => Object.assign({}, ...a)
}
```

**Comment by localvoid** [comment]
@aronallen 

1. Take a look at data structures in highly efficient virtual dom implementations
2. React supports patterns like this: `<A />{condition && <B />}<C />` and in React it will work as anyone is expecting. Many other virtual dom libraries in cases like this will destroy and create `C` when `condition` is changing. Everything is gets way much worse if you add here keys, array nesting, etc, it is a complicated problem.
3. I think you don't understand the attack, it has nothing to do with `innerHTML`.

**Comment by paldepind** [comment]
Closing this as it's not an issue with Snabbdom.
